internal/server/httpapi/exec.go

```diff
@@ -0,0 +1,159 @@
+package httpapi
+
+import (
+	"io"
+	"net/http"
+
+	"github.com/hashicorp/go-hclog"
+	"nhooyr.io/websocket"
+	"nhooyr.io/websocket/wspb"
+
+	"github.com/hashicorp/waypoint/internal/clicontext"
+	pb "github.com/hashicorp/waypoint/internal/server/gen"
+	"github.com/hashicorp/waypoint/internal/serverclient"
+	"github.com/hashicorp/waypoint/internal/serverconfig"
+)
+
+// HandleExec handles the `waypoint exec` websocket API. This works by
+// connecting back to our own local gRPC server.
+func HandleExec(addr string, tls bool) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		log := hclog.FromContext(ctx)
+		log.SetLevel(hclog.Trace)
+
+		// Get our authorization token
+		token := r.URL.Query().Get("token")
+		if token == "" {
+			http.Error(w, "no token provided", 403)
+			return
+		}
+
+		// Accept our websocket connection.
+		c, err := websocket.Accept(w, r, &websocket.AcceptOptions{
+			// We allow connections from anywhere so that the UI can be on another
+			// host and so that we work with proxying. We aren't too worried about
+			// CSRF due to the nature of Waypoint servers and the underlying
+			// protocols and auth tokens, but one day it'd be nice to be stricter
+			// here and accept only known UI hosts (we don't even have this
+			// config at the time of writing).
+			InsecureSkipVerify: true,
+		})
+		if err != nil {
+			log.Info("error accepting websocket connection", "err", err)
```

Comment:
Should this be `log.Error`?

Reply:
Nah, because there are various reasons triggered by the user this can fail. So operators needn't be concerned. This is user-triggered more than anything.